File: litho-core/src/main/java/com/facebook/litho/LithoRenderUnit.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import static com.facebook.litho.annotations.ImportantForAccessibility.IMPORTANT_FOR_ACCESSIBILITY_NO;
import static com.facebook.litho.annotations.ImportantForAccessibility.IMPORTANT_FOR_ACCESSIBILITY_YES;
import static com.facebook.litho.annotations.ImportantForAccessibility.IMPORTANT_FOR_ACCESSIBILITY_YES_HIDE_DESCENDANTS;

import android.graphics.Rect;
import androidx.annotation.IntDef;
import androidx.annotation.Nullable;
import com.facebook.infer.annotation.Nullsafe;
import com.facebook.rendercore.MountItem;
import com.facebook.rendercore.RenderTreeNode;
import com.facebook.rendercore.RenderUnit;
import com.facebook.rendercore.transitions.TransitionRenderUnit;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Nullsafe(Nullsafe.Mode.LOCAL)
public abstract class LithoRenderUnit extends RenderUnit<Object> implements TransitionRenderUnit {

  public static final int STATE_UNKNOWN = 0;
  public static final int STATE_UPDATED = 1;
  public static final int STATE_DIRTY = 2;

  static final int LAYOUT_FLAG_DUPLICATE_PARENT_STATE = 1 << 0;
  static final int LAYOUT_FLAG_DISABLE_TOUCHABLE = 1 << 1;
  static final int LAYOUT_FLAG_MATCH_HOST_BOUNDS = 1 << 2;
  static final int LAYOUT_FLAG_DRAWABLE_OUTPUTS_DISABLED = 1 << 3;
  static final int LAYOUT_FLAG_DUPLICATE_CHILDREN_STATES = 1 << 4;

  @IntDef({STATE_UPDATED, STATE_UNKNOWN, STATE_DIRTY})
  @Retention(RetentionPolicy.SOURCE)
  public @interface UpdateState {}

  private final @Nullable NodeInfo mNodeInfo;
  private final @Nullable ViewNodeInfo mViewNodeInfo;
  private final @Nullable Rect mTouchBoundsExpansion;
  private @Nullable DebugHierarchy.Node mHierarchy; // TODO: remove
  private final Component mComponent;
  private final int mFlags;

  private final int mImportantForAccessibility;

  private final int mUpdateState;

  protected final long mId;
  protected final @Nullable ComponentContext mContext;

  protected LithoRenderUnit(
      long id,
      final Component component,
      final @Nullable NodeInfo nodeInfo,
      final @Nullable ViewNodeInfo viewNodeInfo,
      final @Nullable Rect touchBoundsExpansion,
      final int flags,
      final int importantForAccessibility,
      final @UpdateState int updateState,
      RenderType renderType,
      @Nullable ComponentContext context) {
    super(renderType);
    this.mContext = context;
    mNodeInfo = nodeInfo;
    mViewNodeInfo = viewNodeInfo;
    mTouchBoundsExpansion = touchBoundsExpansion;
    mComponent = component;
    mFlags = flags;
    mImportantForAccessibility =
        importantForAccessibility == IMPORTANT_FOR_ACCESSIBILITY_YES_HIDE_DESCENDANTS
            ? IMPORTANT_FOR_ACCESSIBILITY_YES // the A11Y prop for descendants has been corrected
            : importantForAccessibility;
    mUpdateState = updateState;
    this.mId = id;
  }

  public @Nullable ComponentContext getComponentContext() {
    return mContext;
  }

  @Override
  public long getId() {
    return mId;
  }

  @Override
  public boolean getMatchHostBounds() {
    return (getFlags() & LAYOUT_FLAG_MATCH_HOST_BOUNDS) != 0;
  }

  Component getComponent() {
    return mComponent;
  }

  int getFlags() {
    return mFlags;
  }

  @Nullable
  DebugHierarchy.Node getHierarchy() {
    return mHierarchy;
  }

  void setHierarchy(@Nullable DebugHierarchy.Node node) {
    mHierarchy = node;
  }

  @Nullable
  NodeInfo getNodeInfo() {
    return mNodeInfo;
  }

  @UpdateState
  public int getUpdateState() {
    return mUpdateState;
  }

  public int getImportantForAccessibility() {
    return mImportantForAccessibility;
  }

  boolean isAccessible() {
    if (mImportantForAccessibility == IMPORTANT_FOR_ACCESSIBILITY_NO) {
      return false;
    }

    return (mNodeInfo != null && mNodeInfo.needsAccessibilityDelegate())
        || (mComponent instanceof SpecGeneratedComponent
            && ((SpecGeneratedComponent) mComponent).implementsAccessibility());
  }

  @Nullable
  ViewNodeInfo getViewNodeInfo() {
    return mViewNodeInfo;
  }

  @Nullable
  public Rect getTouchBoundsExpansion() {
    return mTouchBoundsExpansion;
  }

  static @Nullable ComponentContext getComponentContext(MountItem item) {
    return ((LithoRenderUnit) item.getRenderTreeNode().getRenderUnit()).getComponentContext();
  }

  static @Nullable ComponentContext getComponentContext(RenderTreeNode node) {
    return ((LithoRenderUnit) node.getRenderUnit()).getComponentContext();
  }

  static @Nullable ComponentContext getComponentContext(LithoRenderUnit unit) {
    return unit.getComponentContext();
  }

  static LithoRenderUnit getRenderUnit(MountItem item) {
    return getRenderUnit(item.getRenderTreeNode());
  }

  static LithoRenderUnit getRenderUnit(RenderTreeNode node) {
    return (LithoRenderUnit) node.getRenderUnit();
  }

  static boolean isDuplicateParentState(int flags) {
    return (flags & LAYOUT_FLAG_DUPLICATE_PARENT_STATE) == LAYOUT_FLAG_DUPLICATE_PARENT_STATE;
  }

  static boolean isDuplicateChildrenStates(int flags) {
    return (flags & LAYOUT_FLAG_DUPLICATE_CHILDREN_STATES) == LAYOUT_FLAG_DUPLICATE_CHILDREN_STATES;
  }

  static boolean isTouchableDisabled(int flags) {
    return (flags & LAYOUT_FLAG_DISABLE_TOUCHABLE) == LAYOUT_FLAG_DISABLE_TOUCHABLE;
  }

  static boolean areDrawableOutputsDisabled(int flags) {
    return (flags & LAYOUT_FLAG_DRAWABLE_OUTPUTS_DISABLED) != 0;
  }

  public static boolean isMountableView(RenderUnit unit) {
    return unit.getRenderType() == RenderType.VIEW;
  }
}

File: litho-core/src/main/java/com/facebook/litho/LithoRenderUnit.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.util.SparseArray
import com.facebook.litho.annotations.ImportantForAccessibility
import com.facebook.rendercore.MountItem
import com.facebook.rendercore.RenderTreeNode
import com.facebook.rendercore.RenderUnit
import com.facebook.rendercore.transitions.TransitionRenderUnit

abstract class LithoRenderUnit
protected constructor(
    private val _id: Long,
    val component: Component,
    @field:JvmField val commonDynamicProps: SparseArray<DynamicValue<Any?>>?,
    val nodeInfo: NodeInfo?,
    val flags: Int,
    importantForAccessibility: Int,
    renderType: RenderType?,
    @field:JvmField val componentContext: ComponentContext?,
    private val debugKey: String?
) : RenderUnit<Any?>(renderType), TransitionRenderUnit {

  override fun getId(): Long = _id

  override fun getDebugKey(): String = debugKey ?: componentContext?.globalKey ?: _id.toString()

  // TODO: remove
  var hierarchy: DebugHierarchy.Node? = null

  // the A11Y prop for descendants has been corrected
  val importantForAccessibility: Int =
      if (importantForAccessibility ==
          ImportantForAccessibility.IMPORTANT_FOR_ACCESSIBILITY_YES_HIDE_DESCENDANTS) {
        ImportantForAccessibility.IMPORTANT_FOR_ACCESSIBILITY_YES
      } else {
        importantForAccessibility
      }

  override fun getMatchHostBounds(): Boolean = flags and LAYOUT_FLAG_MATCH_HOST_BOUNDS != 0

  val contentDescription: CharSequence?
    get() = nodeInfo?.contentDescription

  val isAccessible: Boolean
    get() {
      if (importantForAccessibility == ImportantForAccessibility.IMPORTANT_FOR_ACCESSIBILITY_NO) {
        return false
      }

      return (this.nodeInfo != null && nodeInfo.needsAccessibilityDelegate()) ||
          (this.component is SpecGeneratedComponent && this.component.implementsAccessibility())
    }

  companion object {
    const val LAYOUT_FLAG_DUPLICATE_PARENT_STATE = 1 shl 0
    const val LAYOUT_FLAG_DISABLE_TOUCHABLE = 1 shl 1
    const val LAYOUT_FLAG_MATCH_HOST_BOUNDS = 1 shl 2
    const val LAYOUT_FLAG_DRAWABLE_OUTPUTS_DISABLED = 1 shl 3
    const val LAYOUT_FLAG_DUPLICATE_CHILDREN_STATES = 1 shl 4
    const val LAYOUT_FLAG_HAS_TOUCH_EVENT_HANDLERS = 1 shl 5

    @JvmStatic
    fun getComponentContext(item: MountItem): ComponentContext? =
        (item.renderTreeNode.renderUnit as LithoRenderUnit).componentContext

    @JvmStatic
    fun getComponentContext(node: RenderTreeNode): ComponentContext? =
        (node.renderUnit as LithoRenderUnit).componentContext

    @JvmStatic
    fun getComponentContext(unit: LithoRenderUnit): ComponentContext? = unit.componentContext

    @JvmStatic
    fun getRenderUnit(item: MountItem): LithoRenderUnit = getRenderUnit(item.renderTreeNode)

    @JvmStatic
    fun getRenderUnit(node: RenderTreeNode): LithoRenderUnit = node.renderUnit as LithoRenderUnit

    @JvmStatic
    fun isDuplicateParentState(flags: Int): Boolean =
        flags and LAYOUT_FLAG_DUPLICATE_PARENT_STATE == LAYOUT_FLAG_DUPLICATE_PARENT_STATE

    @JvmStatic
    fun hasTouchEventHandlers(flags: Int): Boolean =
        flags and LAYOUT_FLAG_HAS_TOUCH_EVENT_HANDLERS == LAYOUT_FLAG_HAS_TOUCH_EVENT_HANDLERS

    @JvmStatic
    fun isDuplicateChildrenStates(flags: Int): Boolean =
        flags and LAYOUT_FLAG_DUPLICATE_CHILDREN_STATES == LAYOUT_FLAG_DUPLICATE_CHILDREN_STATES

    @JvmStatic
    fun isTouchableDisabled(flags: Int): Boolean =
        flags and LAYOUT_FLAG_DISABLE_TOUCHABLE == LAYOUT_FLAG_DISABLE_TOUCHABLE

    @JvmStatic
    fun areDrawableOutputsDisabled(flags: Int): Boolean =
        flags and LAYOUT_FLAG_DRAWABLE_OUTPUTS_DISABLED != 0

    @JvmStatic
    fun isMountableView(unit: RenderUnit<*>): Boolean = unit.renderType == RenderType.VIEW
  }
}
